src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -354,6 +355,18 @@ class RsErrorAnnotator : RsAnnotatorBase(), HighlightRangeExtension {
         }
     }
 
+    // E0118: Can impl only `struct`s, `enum`s and `union`s (when not implementing a trait)
+    private fun checkImplForNonAdtError(holder: AnnotationHolder, impl: RsImplItem) {
+        if (impl.`for` != null) return
+        val typeRef = impl.typeReference ?: return
+        if (typeRef.traitType != null) return
+        val type = typeRef.type
+        if (type !is TyAdt && type != TyUnknown) {
```

Comment:
There is an exception for primitive types: it's possible to make a single impl for primitive type with corresponding lang item, e.g. `#[lang = "u8"] impl u8 {}`. Such impls exist in `libcore`, but it's possible to make them in project in `#![no_core]` mode.

Reply:
Ooh, that's a good catch! I didn't really know about it. Thanks! :)

Reply:
I just found that it's also legal to implement trait objects:
```rust
trait Bar {}
impl dyn Bar {
    fn foo(&self) {}
}
```

Reply:
I believe that's already covered and won't report an error due to this bit of code: `if (typeRef.traitType != null) return`. I will double check just to be sure though

Reply:
`typeRef.traitType != null` is about
```rust
impl Foo for Bar {}
    //^ this trait
```
but not for the case of trait object impl